src/Containers/Microsoft.NET.Build.Containers/Layer.cs

```diff
@@ -44,111 +45,78 @@ public static Layer FromDescriptor(Descriptor descriptor)
         }
 
-
         string tempTarballPath = ContentStore.GetTempFile();
         using (FileStream fs = File.Create(tempTarballPath))
         {
             using (HashDigestGZipStream gz = new(fs, leaveOpen: true))
             {
                 using (TarWriter writer = new(gz, TarEntryFormat.Pax, leaveOpen: true))
                 {
+                    // Write an entry for the application directory.
+                    WriteTarEntryForFile(writer, new DirectoryInfo(directory), containerPath, entryAttributes);
+
+                    // Write entries for the application directory contents.
+                    var fileList = new FileSystemEnumerable<(FileSystemInfo file, string containerPath)>(
+                                directory: directory,
+                                transform: (ref FileSystemEntry entry) =>
+                                {
+                                    FileSystemInfo fsi = entry.ToFileSystemInfo();
+                                    string relativePath = Path.GetRelativePath(directory, fsi.FullName);
+                                    if (OperatingSystem.IsWindows())
+                                    {
+                                        // Use only '/' directory separators.
+                                        relativePath = relativePath.Replace('\\', '/');
+                                    }
+                                    return (fsi, $"{containerPath}/{relativePath}");
+                                },
+                                options: new EnumerationOptions()
+                                {
+                                    RecurseSubdirectories = true
+                                });
                     foreach (var item in fileList)
                     {
-                        // Docker treats a COPY instruction that copies to a path like `/app` by
-                        // including `app/` as a directory, with no leading slash. Emulate that here.
-                        string containerPath = SanitizeContainerPath(item.containerPath);
-
-                        EnsureDirectoryEntries(writer, containerPath.Split(PathSeparators));
-
-                        using var fileStream = File.OpenRead(item.path);
-                        var entry = CreateTarEntry(TarEntryType.RegularFile, containerPath);
-                        entry.DataStream = fileStream;
-
-                        writer.WriteEntry(entry);
+                        WriteTarEntryForFile(writer, item.file, item.containerPath, entryAttributes);
                     }
 
                     // Windows layers need a Hives folder, we do not need to create any Registry Hive deltas inside
                     if (isWindowsLayer)
                     {
-                        var entry = CreateTarEntry(TarEntryType.Directory, "Hives/");
+                        var entry = new PaxTarEntry(TarEntryType.Directory, "Hives/", entryAttributes);
```

Comment:
I guess the trailing slash may not be needed here, but it doesn't cause any harm either.